ETT-61 Exclude */man items from newyear.pl

moseshll opened the pull request:
- Add `NewYear` module with tests for `are_rights_in_scope` extracted and expanded from what is in bin/newyear.pl
  - I have checked the items listed in the tests, which are comprehensive, with KH to get confirmation this is the correct list.
- Add `CRMS::NewYear` method `choose_rights_prediction`
  - This is intended to replace one of the nastiest blocks of code encountered in `bin/newyear.pl` around lines 198-215 and duplicated at 359-376.
  - Much depends on this being comprehensible and working as intended, so this is an attempt to get a better handle on it. Currently this new method is not called anywhere. I would like to do an A/B comparison between the two methods of comparing PDD rights predictions.

KH and I have examined and discussed both methods in the new module and are in agreement that what is here seems right as long as nothing peculiar emerges in the A/B comparison (which should be done by Nov 7 -- it's a time consuming script to run twice). [Edit: see below on result of A/B]

I am now convinced `choose_rights_prediction` should be wired into `newyear.pl` as part of this branch.

moseshll committed:
ETT-61 Exclude */man items from newyear.pl
- Add `NewYear` module with tests for `are_rights_in_scope` extracted and expanded from what is in bin/newyear.pl
- TODO: check the items listed in the tests, which are comprehensive, with KH to get confirmation this is the correct list

moseshll committed:
Add `CRMS::NewYear` method `choose_rights_prediction`

This is intended to replace one of the nastiest blocks of code encountered in `bin/newyear.pl`
around lines 198-215 and duplicated at 359-376. Much depends on this being comprehensible
and working as intended, so this is an attempt to get a better handle on it.

Currently this new method is not swapped in. I would like to do an A/B comparison between the
two methods of comparing PDD rights predictions.

moseshll committed:
Flag two unused variables

moseshll committed:
Clarify comment

moseshll committed:
Use standard GetMetadata block for all projects
 - Make sure to call ClearErrors on the CRMS object, otherwise a catalog error can become sticky and pollute the output
 - Only log metadata failure if verbose

aelkiss commented on lib/CRMS/NewYear.pm:
```diff
@@ -0,0 +1,116 @@
+
+# Are the passed-in current rights "attribute" and "reason" names in scope for
+# PDD rollover? This excludes pd and CC items, because there's nothing more to be done.
+# It also excludes */{con, del, man, pvt, supp} items, because CRMS can't override them
```
On the one hand I feel like pd rights should be able to override CC rights (since once an item is public domain, you can't really limit what people do with it based on a CC license?), but on the other hand we provide *more* access for CC items I think since we can allow full download to non-logged-in users for Google-scanned items that are CC licensed. What a mess.

moseshll replied:
Discussed with KH (and added to ticket). Definitely out of scope for this ticket but useful to consider (I was not aware of this).

aelkiss reviewed (commented):
I wish this process didn't need to exist, but given that it does, this looks fine to me so far:
* the documentation in `CRMS::NewYear` makes sense, and the comments are helpful given the opacity of the logic in `choose_rights_prediction`
* it appears the tests cover the cases discussed in the documentation
* `are_rights_in_scope` makes sense

moseshll committed:
- Swap in CRMS::NewYear::choose_rights_prediction for existing code after successful A/B test.
- Pass current rights to `SubmitNewYearReview` to eliminate a redundant Rights DB hit.
- Short circuit no-op attempt to add to queue when generating TSV report and remove "message" field from the TSV.

moseshll committed:
Add note about random choice between competing pd/add and pd/exp predictions

moseshll commented:
A/B testing result: no discrepancies between the two versions of the code, except approximately five cases where a random choice was made between pd/add and pd/exp. The randomness exists in both the old code and the new, and is a side effect of using hash keys as a set. KH says "don't care" and I have added a comment in NewYear.pm about the behavior.

aelkiss reviewed (approved):
The new changes to use `choose_rights_prediction` look good to me.
